libunwind changes needed to run code in sgx environment via rust-sgx.

AdrianCX opened the pull request:
Please advise on process as this is the first PR against rust-lang's llvm and I tried to follow instructions as close as possible..

Ticket: https://github.com/fortanix/rust-sgx/issues/174

Description: libunwind changes needed to run code in sgx environment via rust-sgx.

Target that uses this in rust: x86_64-fortanix-unknown-sgx.
Rust-lang PR depends on this - plan is to raise it if this PR is accepted.

Without this change, rust std for this toolchain is forced to use a precompiled library loaded via environment variable.

Code is guarded via defines to enable only if 'RUST_SGX' is present.

Main logic is in libunwind/src/AddressSpace.hpp

We use 6 symbols to figure out where eh_frame / eh_frame_hdr is at runtime when loaded in an SGX enclave. (EH symbols + IMAGE base)
These are set when elf is converged to sgx format by 'fortanix-sgx-tools'.

Original ported changes: https://github.com/llvm/llvm-project/compare/release/5.x...fortanix:release/5.x

cuviper commented:
Is it possible to reframe this in a way that upstream LLVM would accept? We don't generally want to carry extra patches indefinitely. With backports, at least we know they'll go away in a future rebase.

jethrogb commented:
> We don't generally want to carry extra patches indefinitely.

The Rust project effectively already does this by [pointing at a a fork](https://github.com/rust-lang/rust/blob/master/src/ci/docker/dist-various-2/build-x86_64-fortanix-unknown-sgx-toolchain.sh#L12-L13). That fork is based on LLVM version 5. Compared to using that external dependency, this PR is a large step in the right direction of bringing us much closer to upstream.

jethrogb commented on libunwind/src/UnwindRustSgxSnprintf.c:
```diff
@@ -0,0 +1,867 @@
+/*
```
I think this entire file is no longer necessary?

AdrianCX replied:
correct, removed, thank you

jethrogb commented on libunwind/src/libunwind.cpp:
```diff
@@ -288,6 +288,11 @@ void __unw_remove_dynamic_fde(unw_word_t fde) {
 
 _LIBUNWIND_HIDDEN
 bool logAPIs() {
+#ifdef RUST_SGX
```
Looks like these APIs are used only in debug builds, can these changes be avoided?

AdrianCX replied:
removed these changs, we never define the NDEBUG flag so they are never compiled in our use case.

jethrogb commented on libunwind/src/UnwindRustSgx.h:
```diff
@@ -0,0 +1,92 @@
+#define malloc libuw_malloc
+#define free libuw_free
+
+#undef dl_iterate_phdr
```
Is this still necessary?

AdrianCX replied:
removed #undef dl_iterate_phdr, thanks.

jethrogb commented on libunwind/src/UnwindRustSgx.h:
```diff
@@ -0,0 +1,90 @@
+    int __rust_rwlock_unlock(RWLock *rwlock);
+    unsigned char *__rust_c_alloc(size_t, size_t);
+    void __rust_c_dealloc(unsigned char *, size_t, size_t);
+    void __rust_print_err(uint8_t *m, int s);
```
I think this can be conditional on `NDEBUG` now

cuviper commented:
Please note that we've moved to LLVM 10 since https://github.com/rust-lang/rust/pull/67759, so we'll need this PR rebased onto `rustc/10.0-2020-05-05` to reach rust master.

AdrianCX committed:
Changes needed for 'x86_64-fortanix-unknown-sgx' nightly target.

Code is guarded via defines to enable only if 'RUST_SGX' is present.

Main logic is in libunwind/src/AddressSpace.hpp
We use 6 symbols to figure out where eh_frame / eh_frame_hdr is at runtime when loaded in an SGX enclave. (EH symbols + IMAGE base)
These are set by 'fortanix-sgx-tools'.

As notes:
- Target above at the moment uses a pre-compiled libunwind.a from forked repo.
- Goal of these changes is to use official llvm with patch.
- Changes in rust-lang to use this are planned if/when this is accepted.
- Ticket: fortanix/rust-sgx#174
- Original ported changes: llvm/llvm-project@release/5.x...fortanix:release/5.x

AdrianCX commented:
> Please note that we've moved to LLVM 10 since [rust-lang/rust#67759](https://github.com/rust-lang/rust/pull/67759), so we'll need this PR rebased onto `rustc/10.0-2020-05-05` to reach rust master.

Thanks, I updated this PR with required changes and modified target.

cuviper commented:
Ok, I'm going to trust @jethrogb's review for SGX's sake, and this looks harmless to other targets.
